### ISSUE TITLE
Adobe Reader fails to read aloud Vivliostyle output PDF accessibility issue

### ISSUE THREAD

**Issue by danvim** [vivliostyle/vivliostyle.js#1555]
**Describe the bug**
Vivliostyle output PDFs when opened with Adobe Reader just fails to read aloud.

**To Reproduce**
1. Go to Alice in Wonderland Vivliostyle sample with Chrome: https://vivliostyle.org/viewer/#src=https://vivliostyle.github.io/vivliostyle_doc/samples/gutenberg/Alice.html&bookMode=true
2. Print PDF in Vivliostyle viewer.
3. Open the Vivliostyle output PDF in Adobe Reader > View > Read out loud > Activate Read Out Loud
4. Adobe Reader > View > Read out loud > Read to end document then listen to narration

**Expected behavior**
Accessibility features exported by the browser is not lost. Adobe reader reading out text content generated by Chrome/Chromium based browsers

**Actual behavior**
Adobe reader reads "Warning: empty page" for all pages. If we just to Ctrl+P directly on Alice in Wonderland HTML document https://vivliostyle.github.io/vivliostyle_doc/samples/gutenberg/Alice.html with Chrome, Adobe Reader is able to read aloud without any issues.

**Reference**
Comparing with paged.js https://pagedjs.org/en/examples/ "Situer le numérique". This pdf can be read aloud by Adobe Reader, albeit each line having a pause in between.

PAC Accessibility Checker: https://pac.pdf-accessibility.org/en/check

## Chrome native Alice in Wonderland

<img width="1235" height="956" alt="Image" src="https://github.com/user-attachments/assets/eefe4e81-6dce-494f-8b69-3154f915e484" />

<img width="1423" height="1572" alt="Image" src="https://github.com/user-attachments/assets/ef6ceef1-fa11-4f2c-80e4-c34bbc3bdb71" />

## Vivliostyle on Chrome Alice in Wonderland

<img width="1903" height="956" alt="Image" src="https://github.com/user-attachments/assets/32e62315-0524-4a79-a4ac-513238229162" />

<img width="1565" height="1836" alt="Image" src="https://github.com/user-attachments/assets/a976e474-c9e9-4cc7-8f84-4cbf52ddc9f0" />

## Suspected issue

Following experiments are done on Vivliostyle Viewer

### Disabling transform zooming styles seems to have fixed it

<img width="2084" height="1431" alt="Image" src="https://github.com/user-attachments/assets/3b4b69b9-c5a9-4bdd-a623-4b224dde4b08" />

### Applying zoom while disabling transform still exports marked content

<img width="2443" height="1149" alt="Image" src="https://github.com/user-attachments/assets/f88953e0-085c-43e9-80f4-73659ec8ce65" />

<img width="961" height="1089" alt="Image" src="https://github.com/user-attachments/assets/8be5c663-1e62-4ca9-8759-baf5acfe67a6" />

**Comment by MurakamiShinyu** [comment]
Thank you for the bug report and your detailed investigation. I found that this issue was introduced in [v2.21.0](https://github.com/vivliostyle/vivliostyle.js/releases/tag/v2.21.0) with the addition of the feature for enabling very thin border output, which relies on CSS zoom and transform properties. Disabling this feature (by specifying `pixelRatio=0` to the Vivliostyle Viewer's URL parameter) can avoid the issue, e.g.,
https://vivliostyle.org/viewer/#src=https://vivliostyle.github.io/vivliostyle_doc/samples/gutenberg/Alice.html&pixelRatio=0

See the pull request of the feature that introduced this issue:
- https://github.com/vivliostyle/vivliostyle.js/pull/1079

This feature is enabled by default (`pixelRatio=8`, the minimum border width is 1/8 px) and causes the issue to occur. I will consider disabling this feature by default if we cannot find a better solution that avoids the Chromium issue ([Cannot print with border-width of less than 1pt](https://issues.chromium.org/issues/40149824)) without causing this accessibility issue.

**Comment by danvim** [comment]
Thanks for the recommendation. I also came across that Chromium issue. Quite the shame. I tried on different browsers and somehow Safari is the only one that can print thin lines. But it can't output Vivliostyle/paged.js PDF properly.

On the other hand, I'm not sure if your team monitors this tool, but here are some suggestions based on its output report, adding ARIA attributes and roles may help with the output. e.g.

-  Adding `role="presentation"` on page wrappers and everything outside seems to help with simplifying the output tree structure. The effect would be unwrapping the PDF "Div" tag in my experiment.
-  Adding default roles for margin boxes so they are tagged in the output (e.g. page number). This seems to be one gap that can be filled to achieve better PDF/UA output. Suggestion: `role="complementary"`? `<aside>`? But not sure if this would affect running elements. The book title on each page and page number are currently marked as PDF "P" tag
- Not much of a suggestion, but I also notice the ordering of the elements from the DOM translates directly to the ordering of the PDF tagged content.

**Comment by MurakamiShinyu** [comment]
Thank you for your suggestions for improving accessibility in the PDF output. I am going to implement these suggestions in Vivliostyle.js.

I found a workaround for the issue with `transform: scale(N)` causing problems in Chromium when generating PDFs.
The `scale(N)` is equivalent to `matrix(N, 0, 0, N, 0, 0)`. I tested this and it causes the same problem.
However, when the 2nd or 3rd parameters of the matrix are non-zero, it does not cause the problem.
So we can use `transform: matrix(N, 0, 5e-324, N, 0, 0)` as a workaround. The `5e-324` is the smallest non-zero number that browsers can handle, so it effectively acts as `0` without causing the issue.

I am going to implement this workaround soon.

**Comment by MurakamiShinyu** [comment]
Fixed in PR #1559.

Test:
- https://vivliostyle-git-fix-issue1555-vivliostyle.vercel.app/#src=https://vivliostyle.github.io/vivliostyle_doc/samples/gutenberg/Alice.html

@danvim Thanks for your help. Could you check this fix, please!

**Comment by danvim** [comment]
> Fixed in PR [#1559](https://github.com/vivliostyle/vivliostyle.js/pull/1559).
> 
> Test:
> 
> * https://vivliostyle-git-fix-issue1555-vivliostyle.vercel.app/#src=https://vivliostyle.github.io/vivliostyle_doc/samples/gutenberg/Alice.html
> 
> [@danvim](https://github.com/danvim) Thanks for your help. Could you check this fix, please!

This test works. Thanks for the quick fix. Here are the screenshots for comparison. I've included the rest of the screens just for completeness. 

The only gap in the PDF/UA validity seems to me like the PDF metadata and missing bookmarks, which can be post-processed by the user, and related to the original content itself (text contrast, li structure, and link structure warning complaint which I do not understand).

## Thin border test (1pt vs 0.5pt vs 0.2pt)

https://vivliostyle-git-fix-issue1555-vivliostyle.vercel.app/#src=data:text/html,%3Cdiv%20style=%22border:%201pt%20solid%20red;width:%2010pt;%20height:%2010pt;%22%3E%3C/div%3E%3Cdiv%20style=%22border:%200.5pt%20solid%20red;width:%2010pt;%20height:%2010pt;%22%3E%3C/div%3E%3Cdiv%20style=%22border:%200.2pt%20solid%20red;width:%2010pt;%20height:%2010pt;%22%3E%3C/div%3E&style=data:,/*%3Cviewer%3E*/%0A@page%20%7B%20size:%20a4;%20%7D%0A/*%3C/viewer%3E*/

<img width="645" height="613" alt="Image" src="https://github.com/user-attachments/assets/73cc9a15-3f86-4d3d-913a-8425731f4ea8" />

## Alice test files

- [Alice's Adventures in Wonderland, by Lewis Carroll fixed.pdf](https://github.com/user-attachments/files/22368108/Alice.s.Adventures.in.Wonderland.by.Lewis.Carroll.fixed.pdf)
- [Alice's Adventures in Wonderland, by Lewis Carroll chrome native.pdf](https://github.com/user-attachments/files/22368107/Alice.s.Adventures.in.Wonderland.by.Lewis.Carroll.chrome.native.pdf)

There are many screenshots, but they get very long so I've collapsed them underneath the PDF/UA screen for each file.

## Fix test

<img width="964" height="1246" alt="Image" src="https://github.com/user-attachments/assets/091f4de7-559b-45f4-acb3-5f971b35c323" />

<details><summary><h3>Other screenshots</h3></summary>

<img width="964" height="1246" alt="Image" src="https://github.com/user-attachments/assets/07bb0a83-a9d9-4671-945a-01b54158692b" />

<img width="962" height="1316" alt="Image" src="https://github.com/user-attachments/assets/daf78518-e694-46ac-95ad-dd97cdb63464" />

<img width="1437" height="1906" alt="Image" src="https://github.com/user-attachments/assets/a2e7d1a7-01b5-4fb5-afd2-099137877df8" />

<img width="1943" height="902" alt="Image" src="https://github.com/user-attachments/assets/a1b1d97e-b443-46d6-b306-b4ef8d58f7c4" />

<img width="2423" height="919" alt="Image" src="https://github.com/user-attachments/assets/e3675603-c248-402e-939f-0d3bb98b0b79" />

<img width="1973" height="1076" alt="Image" src="https://github.com/user-attachments/assets/cb95f0ec-2836-42d7-bdb9-d9cb5d0ee8e6" />

<img width="1943" height="902" alt="Image" src="https://github.com/user-attachments/assets/43b7182b-7ba9-40c1-9d74-d62fc920ca95" />

<img width="1480" height="1449" alt="Image" src="https://github.com/user-attachments/assets/eca9b3ef-1e0a-4b8d-9b83-4bbc465f5d15" />

</details>

## Chrome native

<img width="962" height="1316" alt="Image" src="https://github.com/user-attachments/assets/3c5c6921-758e-483b-a03f-5ae3d8005fcb" />

<details><summary><h3>Other screenshots</h3></summary>

<img width="962" height="1316" alt="Image" src="https://github.com/user-attachments/assets/2247c911-da8f-48fa-8f89-dd45538b8f9b" />

<img width="962" height="1316" alt="Image" src="https://github.com/user-attachments/assets/033ca7dc-1637-4fe8-b8e2-d70fe7579aca" />

<img width="1256" height="1701" alt="Image" src="https://github.com/user-attachments/assets/4cc5dc59-a9dc-4615-8fcf-87471cd08906" />

<img width="1943" height="902" alt="Image" src="https://github.com/user-attachments/assets/aaf35690-7f9b-4082-9e62-66c4aaee1af1" />

<img width="1943" height="902" alt="Image" src="https://github.com/user-attachments/assets/e82e2b3d-4ff9-4182-8d52-2449adee1459" />

<img width="1943" height="902" alt="Image" src="https://github.com/user-attachments/assets/e8a2ebf4-8e49-4b20-b25c-1a4dc53060d4" />

</details>


## Apostrophe "s" tripping up MS Edge read aloud

Edge reads this header in the fixed PDF, as "Alice... Es adventure in wonderland". I wonder if it's related to this `<viv-ts-close>` node being converted into a `NonStruct` PDF tagged element. Edge also reads the page number now. When using Edge to read the Chrome native produced PDF, it skips the header and footer (probably precisely since they were untagged like initially). I think more testing and research is probably needed, but probably for another issue.

<img width="1480" height="1449" alt="Image" src="https://github.com/user-attachments/assets/6b708f4a-c835-4a49-b255-c8a7124c9c66" />

<img width="2361" height="1595" alt="Image" src="https://github.com/user-attachments/assets/aebaf6b4-8550-4d27-9aa7-488713721fa8" />

Here are some resources I found that talks about marking headers and footers as `Artifact`s (probably corresponding to `aria-hidden`):

- https://www.w3.org/TR/WCAG20-TECHS/PDF14.html
- https://www.davidmacd.com/blog/pdf-headers-footers.html

**Comment by MurakamiShinyu** [comment]
> ## Apostrophe "s" tripping up MS Edge read aloud
> 
> Edge reads this header in the fixed PDF, as "Alice... Es adventure in wonderland". I wonder if it's related to this `<viv-ts-close>` node being converted into a `NonStruct` PDF tagged element.

Thanks for finding the problem. This is a side effect of the implementation of the CSS hanging-punctuation and text-spacing properties in Vivliostyle.js. We need to improve this implementation, so we will treat it as a separate issue.

> Here are some resources I found that talks about marking headers and footers as `Artifact`s (probably corresponding to `aria-hidden`):
> 
> - https://www.w3.org/TR/WCAG20-TECHS/PDF14.html
> - https://www.davidmacd.com/blog/pdf-headers-footers.html

Thanks for the useful information. I added `role="complementary"` to the page margin boxes as you suggested, but I was unsure whether to also specify `aria-hidden="true"`. When reading aloud a text that spans multiple pages, it may be annoying if the header content and page numbers are read aloud at the page breaks, but there may also be cases where they should be read aloud. We will consider this as a separate issue.

**Comment by MurakamiShinyu** [comment]
I reported the issue to the Chromium Issue Tracker:
- [PDF Accessibility - untagged text is generated if transform:scale() is used](https://issues.chromium.org/issues/447444703)
